runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/endpoints/Endpoint.kt

```diff
@@ -27,9 +23,52 @@ import aws.smithy.kotlin.runtime.http.Url
  * is expected to be mutable and the client cannot modify the endpoint correctly, the operation
  * will likely fail.
  */
-public data class Endpoint(
-    public val uri: Url,
-    public val isHostnameImmutable: Boolean = false,
-) {
+public class Endpoint {
+    /**
+     * The base URL endpoint clients will use to make API calls to e.g. "api.myservice.com".
+     * NOTE: Only `scheme`, `port`, `host` `path`, and `parameters` are valid. Other URL elements are ignored.
+     */
+    public val uri: Url
+
+    /**
+     * Flag indicating whether the hostname can be modified by the SDK client.
+     */
+    public val isHostnameImmutable: Boolean
+
+    /**
+     * A map of additional HTTP headers to be set when making calls against this endpoint.
+     */
+    public val headers: Map<String, List<String>>
+
+    /**
+     * A grab-bag property map of endpoint attributes. The values here are only set when the endpoint returned from
+     * evaluating a ruleset.
+     */
+    @InternalApi
+    public val attributes: Map<String, Any>
+
+    public constructor(
+        uri: Url,
+        isHostnameImmutable: Boolean = false,
+        headers: Map<String, List<String>> = mapOf(),
+    ) {
+        this.uri = uri
+        this.isHostnameImmutable = isHostnameImmutable
+        this.headers = headers
+        this.attributes = mapOf()
+    }
+
     public constructor(uri: String) : this(Url.parse(uri))
+    @InternalApi
+    public constructor(
```

Comment:
why so many constructors? Seems like a single constructor that defaults hostnameImmutable/headers/attributes would suffice

Reply:
Fixed by reverting to ADT.